flow/record/fieldtypes/__init__.py

```diff
@@ -542,23 +542,64 @@ def _unpack(cls, data):
+    return isinstance(path, pathlib.PurePath) and (path._flavour.sep == "\\" or path._flavour.altsep == "\\")
+
+
 class path(pathlib.PurePath, FieldType):
     def __new__(cls, *args):
+        # This is modelled after pathlib.PurePath's __new__(), which means you
+        # will never get an instance of path, only instances of either
+        # posix_path or windows_path.
         if cls is path:
+            # By default, path will behave differently on windows and posix
+            # systems, similarly as pathlib.PurePath does.
             cls = windows_path if os.name == "nt" else posix_path
-            if len(args) > 0:
-                path_ = args[0]
-                if isinstance(path_, pathlib.PureWindowsPath):
+
+            # Try to determine the path behaviour based on the first path part
+            # in args that has pathlib.PurePath traits.
+            for path_part in args:
+                print(f"{path_part=} {type(path_part)=}")
+                if isinstance(path_part, pathlib.PureWindowsPath):
+                    cls = windows_path
+                    # The (string) representation of a pathlib.PureWindowsPath is not round trip equivalent if a
+                    # path starts with a \ or / followed by a drive letter, e.g.: \C:\...
+                    # Meaning:
+                    #
+                    # str(PureWindowsPath(r"\C:\WINDOWS/Temp")) !=
+                    # str(PureWindowsPath(PureWindowsPath(r"\C:\WINDOWS/Temp"))),
+                    #
+                    # repr(PureWindowsPath(r"\C:\WINDOWS/Temp")) !=
+                    # repr(PureWindowsPath(PureWindowsPath(r"\C:\WINDOWS/Temp"))),
+                    #
+                    # This would be the case though when using PurePosixPath instead.
+                    #
+                    # This construction works around that by converting all path parts that are
+                    # pathlib.PureWindowsPath descendants to strings first.
+                    args = (str(arg) if isinstance(arg, pathlib.PureWindowsPath) else arg for arg in args)
+                elif isinstance(path_part, pathlib.PurePosixPath):
+                    cls = posix_path
+                elif _is_windowslike_path(path_part):
+                    # This handles any custom PurePath based implementations that have a windows
+                    # like path separator (\).
                     cls = windows_path
-                    args = [path_.as_posix()]
-                elif isinstance(path_, pathlib.PurePosixPath):
+                    args = (str(arg) if _is_windowslike_path(arg) else arg for arg in args)
+                elif _is_posixlike_path(path_part):
+                    # This handles any custom PurePath based implementations that don't have a
+                    # windows like path separator (\).
                     cls = posix_path
-                    args = [path_.as_posix()]
+                else:
+                    continue
```

Comment:
What happens here? Is this a realistic case, i.e. we don't know? Why make a loop anyway instead of just picking the first element from args?

Reply:
The type of path is determined by the first path part from which such a thing can be inferred. In the case of this implementation that means being a subclass of `pathlib.PurePath`.

As path parts are also allowed to be plain `str` types (or `PathLike`, which amounts to being str typed), we have to iterate over the parts until we find the first part that has a type we are looking for.  The only Python types that can make explicit how paths are formatted are `PurePath` subclasses. So we have to iterate and discard all path parts that are not a `PurePath` subclass.

The native `PurePath` doesn't do any inference at all, as it will always return a path interpreted according to how the local system defines paths. If the user doesn't want that behavior, they always have to explicitly use the correct `PurePath` subclass.

For the flow.record `path` class we do some form of auto detection for convenience, as flow.record is often used in systems that deal with paths which are not formatted according to the local system specifications.